### PR TITLE
Listview added intially

### PR DESCRIPTION
**Pull request by Moeez1999**

### CODE REVIEW

File: lib/layout/box_layout.dart [hunk 1]
```diff
@@ -154,7 +164,9 @@ class BoxLayoutState extends WidgetState<BoxLayout> with TemplatedWidgetState {
       children = updatedChildren;
     }
 
-    Widget boxWidget = _buildBoxWidget(children);
+    Widget boxWidget = templatedChildren != null && widget is ListView
+        ? _buildListView(children)
```
Comment (vusters) [inline lib/layout/box_layout.dart:168]: if else is not ideal as ListView is very different than BoxWidget. Can you extract the barebone (didChangeDependencies(), dispose()) into a base class say "WidgetListState", then have both BoxLayoutState and a new ListViewState extend it?

Comment (vusters) [reply lib/layout/box_layout.dart:168]: that way ListViewState is clean, and knows nothing about the Column/Row/...

File: lib/layout/box_layout.dart [hunk 2]
```diff
@@ -200,6 +212,16 @@ class BoxLayoutState extends WidgetState<BoxLayout> with TemplatedWidgetState {
 
   }
 
+  Widget _buildListView(List<Widget> listViewChildren) {
+    return flutter.ListView.builder(
+        scrollDirection: Axis.vertical,
+        itemCount: listViewChildren.length,
+        shrinkWrap: true,
+        itemBuilder: (BuildContext context, int index) {
```
Comment (vusters) [inline lib/layout/box_layout.dart:220]: Can you run a test of say 100 widgets. All 100 widgets will be constructed and called their initState(), but only the visible one in the ListView will call their buildWidget() method?

Comment (vusters) [reply lib/layout/box_layout.dart:220]: I think it should work like that but if it doesn't we have to do proper paging.

File: lib/page_model.dart [hunk 3]
```diff
@@ -165,7 +165,7 @@ class PageModel {
   WidgetModel buildRootModel(YamlMap viewMap, Map<String, dynamic>? customViewDefinitions) {
     WidgetModel? rootModel = getRootModel(viewMap, customViewDefinitions!);
     if (rootModel != null) {
-      if (![Column.type, Row.type, Flex.type, EnsembleStack.type, AppScroller.type].contains(rootModel.type)) {
+      if (![Column.type, Row.type, Flex.type, EnsembleStack.type, AppScroller.type, ListView.type].contains(rootModel.type)) {
```
Comment (vusters) [inline lib/page_model.dart:168]: remove this. ListView doesn't need to be the root widget. The rootwidget is meant as a wrapper to include other widgets.

File: lib/layout/list_view.dart [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  static const type = 'ListView';
+  ListView({Key? key}) : super(key: key);
+
+  @override
+  bool isVertical() {
+    return true;
+  }
+}
+
+abstract class BoxLayout extends StatefulWidget
```
Comment (vusters) [inline lib/layout/list_view.dart:21]: I don't understand. Is this a duplicate of the original BoxLayout? Did you mean to remove the original one?

File: lib/layout/list_view.dart [hunk 5]
```diff
@@ -8,22 +9,17 @@ import 'package:flutter/cupertino.dart';
 import 'package:flutter/material.dart' as flutter;
 
 // ignore: must_be_immutable
-class ListView extends BoxLayout {
+class ListView extends ListViewWidget {
```
Comment (vusters) [inline lib/layout/list_view.dart:12]: I didn't catch this before. Why have both ListView and ListViewWidget when they are not used by anything else? Please combine them

**Review by vusters — COMMENTED** [review]
- Rename onTap to onItemTap. We want to know when each item is tapped, not the whole ListView.

**Comment by vusters** [comment]
Go ahead and pull from main and merge your existing code. It is out of date. Once done, you can implement my changes suggested here

**Comment by vusters** [comment]
Hey @Moeez1999 Please read over the specs carefully https://github.com/orgs/EnsembleUI/projects/2/views/1?pane=issue&itemId=15643685. You don't have most of those implemented. For clarity I removed the boxLayout styling and list them all out for you. You may merge the code then add all of them in.

File: lib/layout/list_view.dart [hunk 6]
```diff
@@ -106,22 +105,43 @@ class ListViewWidgetState extends WidgetState<ListViewWidget> with TemplatedWidg
...
+            shrinkWrap: true,
+            itemBuilder: (BuildContext context, int index) {
+              return GestureDetector(
+                  onTap: widget._controller.onItemTap == null ? null : () =>
+                      ScreenController().executeAction(context, widget._controller.onItemTap!),
+                  child: children[index]);
+            }),
+      ),);
+  }
+  BoxDecoration _buildBoxDecoration() {
```
Comment (vusters) [inline lib/layout/list_view.dart:131]: did you copy this from some other widgets? can you re-use it as function instead?

**Review by vusters — COMMENTED** [review]
Looks good. What about backgroundColor and separator*?

**Review by vusters — APPROVED** [review]

**Comment by kmahmood74** [comment]
@Moeez1999 every commit MUST HAVE THE URL TO THE TICKET as the first comment. This way github automatically links it.

**Pull request merged**